src/mongoc/mongoc-uri.c

```diff
@@ -737,13 +738,13 @@ _mongoc_uri_build_write_concern (mongoc_uri_t *uri) /* IN */
          value = bson_iter_int32 (&iter);
 
          switch (value) {
-         case -1:
-            mongoc_write_concern_set_w (write_concern,
-                                        MONGOC_WRITE_CONCERN_W_ERRORS_IGNORED);
-            break;
-         case 0:
-            mongoc_write_concern_set_w (write_concern,
-                                        MONGOC_WRITE_CONCERN_W_UNACKNOWLEDGED);
+         case MONGOC_WRITE_CONCERN_W_ERRORS_IGNORED:
+         case MONGOC_WRITE_CONCERN_W_UNACKNOWLEDGED:
+            /* Warn on conflict, since write concern will be validated later */
+            if (mongoc_write_concern_get_journal(write_concern)) {
+               MONGOC_WARNING("Journal conflicts with w value [w=%d].", value);
```

Comment:
There's a little bit of a facility in test-libmongoc.c right now (suppress_one_message(), which suppresses N messages based on incrementing a global).

It's a little too broad right now because the suppressions include messages which aren't usually printed (INFO level).  And if we want to use it to verify that warnings are triggered we should probably do something fancier, but that's the facility you're basically looking for.